Change: use precompiled headers to speed up compilation

rubidium42 opened the pull request:
## Motivation / Problem

Compilation is slow; precompiled headers make it faster.


## Description

Enable precompiled headers in cmake for `stdafx.h`, `3rdparty/fmt/format.h` and `table/strings.h`, though that required a bump of minimum cmake version.

```
real    12m2.645s
user    43m25.293s
sys     2m39.714s
```
=>
```
real    9m35.804s
user    33m50.140s
sys     2m19.130s
```


## Limitations

18 years ago it seemed to have caused problems, but compilers have probably improved quite a bit since then.


## Scope questions

Should this PR also include the common STL headers into stdafx.h?
Should stdafx.h (ancient MSVC naming) be renamed to the currently more common pch.h?


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 committed:
Change: use precompiled headers for stdafx.h and 3rdparty/fmt/format.h

TrueBrain commented on src/strgen/strgen.cpp:
```diff
@@ -269,6 +269,7 @@ struct HeaderFileWriter : HeaderWriter, FileWriter {
 		fprintf(this->fh, "/* This file is automatically generated. Do not modify */\n\n");
 		fprintf(this->fh, "#ifndef TABLE_STRINGS_H\n");
 		fprintf(this->fh, "#define TABLE_STRINGS_H\n");
+		fprintf(this->fh, "typedef uint32_t StringID;\n");
```
Why this change? Can't find it in the description either. Feel a bit weird, in this PR?

TrueBrain replied:
ah, found the answer myself, as otherwise you cannot precompile that header. But this feels like the wrong solution? Shouldn't we include some header that defines this type?

rubidium42 replied:
I tried header inclusion, specifically `#include "strings_type.h"` and `#include "../strings_type.h"` (because it's as if it is in the table folder for the normal compile). But both give an error that it can't find the file because it's resolving the name relative to `<user defined path>/generated/table/strings.h`. So slapping a couple of `../`s to it would also not reliably get it to the right path.
And at that point the question is... how much effort are we going to put into something relatively brittle when just adding a single typedef works just fine.

glx22 replied:
Aren't we also trying to replace `typedef` with `using` ?

TrueBrain replied:
The manual suggest to include `strings_type.h` in the precompiled header as well, to resolve that issue.

I do not think it is advised to define `StringID` twice in our source code; that is just asking for trouble. So I agree that a `../` is not a good path, but than dismissing: so we should do a typedef like this, seems too short-sighted :) I am sure we can find something else that is elegant and sweet :)

PeterN replied:
Yeah, as glx said, `using StringID = uint32_t;` if it is needed.

glx22 replied:
Including `strings_type.h` works for me, but I also have the CMakeLists.txt change I suggested

rubidium42 replied:
With glx's suggestion including "strings_type.h" seems to do the trick.

glx22 commented on CMakeLists.txt:
```diff
@@ -243,6 +246,13 @@ if(MSVC)
+    PRIVATE
+    src/stdafx.h
+    src/3rdparty/fmt/format.h
+    "${CMAKE_BINARY_DIR}/generated/table/strings.h"
```
Maybe
```suggestion
    [["table/strings.h"]]
```
?

rubidium42 committed:
Change: disable precompiled headers on Linux (GCC - Dedicated) target to ensure those still build

TrueBrain commented on .github/workflows/ci-build.yml:
```diff
@@ -256,10 +256,23 @@ jobs:
+        - name: Windows (windows-latest / x86) without precompiled headers
+          arch: x64
+          os: windows-latest
+          extra-cmake-parameters: -DCMAKE_DISABLE_PRECOMPILE_HEADERS=true
```
Normally it is `ON` for CMake to enable things. (and not `true`).

TrueBrain commented on .github/workflows/ci-build.yml:
```diff
@@ -256,10 +256,25 @@ jobs:
+        - name: windows-2019 / x86
+          arch: x86
+          os: windows-2019
+        - name: windows-2019 / x64
```
personally, I wouldn't do this `name` entry. Without it, and having the `name: Windows (..)` as it was, saves a field we can make mistakes with.

TrueBrain replied:
owh, no, I see why you did this. Maybe we are better off adding this to the GCC Dedicated run. It is already a special run to check these kinds of things, and saves us from this trouble?

rubidium42 replied:
Though how would you see the difference between the one with and the one without precompiled headers if it's only showing arch and OS?

rubidium42 replied:
> owh, no, I see why you did this. Maybe we are better off adding this to the GCC Dedicated run. It is already a special run to check these kinds of things, and saves us from this trouble?

The reason I choose MSVC is because GCC, and Clang on Linux, will happily accept it in our current codebase to not include `optional` when using `std::optional`, whereas MSVC just fails to build. That's the whole reason I went for MSVC, and the x86 variant is because that's most picky about `size_t` to `int` conversions.

Though indeed making GCC Dedicate the one without precompiled headers is the easiest, but will it catch most of the issues? That's the conundrum I've been thinking about.

TrueBrain replied:
I could argue that nobody is going to run MSVC without precompiled headers (as it is on by default on all new MSVC projects anyway), while I am sure there will be some Linux variants where it doesn't work (looking at you GCC 11) :P

I would go for our sanity, and keep the CI file as simple as we can .. it is already scary complex :)

rubidium42 committed:
Change: include fmt.h C++ headers in stdafx.h

This to prevent compilation issues between runs with and without precompiled
headers. Also remove the headers from the rest of the code base as they are
not needed there anymore, although they do relatively little harm.
